### ISSUE TITLE
Unable to Upload to Garmin 

### ISSUE THREAD

**Issue by 20TomTom** [kochka/WeightLogger#68]
Passwords and everything checked. Garmin has had an update ....

**Comment by lolo406** [comment]
I have the same problem

**Comment by ajayrana1310** [comment]
Same problem.

**Comment by Maarty** [comment]
Garmin recently changed authentication to OAuth, it first did not affect the endpoint for .fit upload, but probably now it does.   
Some inspiration on how to make this work can be found in this lib https://github.com/matin/garth/tree/main

**Comment by rakatika** [comment]
Anyone manage to solve this?

**Comment by Maarty** [comment]
I am not a Java developer, but if someone wants to fix this, basically, following the pattern in this lib https://github.com/matin/garth/tree/main you need to do the following steps:
1. get the correct cookies and "ticket" - already done in `app/src/main/java/org/kochka/android/weightlogger/tools/GarminConnect.java`
2. Get OAuth1 token:
- GET request to https://connectapi.garmin.com/oauth-service/oauth/preauthorized/?ticket={ticket}&login-url=https://sso.garmin.com/sso/embed&accepts-mfa-tokens=true 
- "User-Agent" header =  "com.garmin.android.apps.connectmobile"
- "Authorization" header = `OAuth XXXXXXXXXX ` where XXX is a standard OAuth1 header. Consumer key and consumer secret are here `https://thegarth.s3.amazonaws.com/oauth_consumer.json`
3. Exchange OAuth1 token for OAuth2 token
-  POST request to https://connectapi.garmin.com/oauth-service/oauth/exchange/user/2.0
- "User-Agent" =  "com.garmin.android.apps.connectmobile"
- "Accept-Encoding" = "gzip, deflate"
- "Accept" = "*/*"
- "Authorization" header = `OAuth XXXXXXXXXX ` - standard OAuth1 header with retrieved OAuth1 token from step 2
4. In step 3 you receive an OAuth2 token, which then can be used in all API requests as a Bearer token in the Authentication header

The OAuth2 token has a classic expiration, and it can be refreshed by a refresh token, or regenerated by exchanging the OAuth1 token again (that one has a 1-year expiration).  
Also, in the API url stop using `modern/proxy` in the path and that will work.

**Comment by eniac2000** [comment]
Same problem here

**Comment by mikelikesspam** [comment]
This app has 50k+ downloads on Google Play.  I'd happily pay for an upgraded app that supported 2fa and sync to Garmin.  Surely there are enough people willing to buy a cup of coffee to make this worthwhile for a  new developer to pickup. 

@matin -  I don't know you so forgive me plugging you here, but this seems like a trivial fix for you. Your code is already cited above.

**Comment by matin** [comment]
@kochka I'm pretty terrible at Java, but I'm happy to provide guidance or help in any way. It looks like @Maarty outlined the correct auth flow perfectly. Are you up for making the change or at least reviewing a PR?

**Comment by matin** [comment]
> @matin - I don't know you so forgive me plugging you here, but this seems like a trivial fix for you. Your code is already cited above.

@mikelikesspam no forgiveness required 😄. Always happy to help in any way I can!

**Comment by Syntax753** [comment]
Created an initial draft PR for these changes following your excellent insights @matin !
Still WIP and imagine I need to handle the cookies returned after signin..

Any insights would be fantastic - and happy holidays if this goes in to next year :D

https://github.com/kochka/WeightLogger/pull/69

**Comment by kochka** [comment]
@matin I'm not a Java developer neither, I did this app back in the day to fit my needs 😅 
I'm currently really busy but I'll try to help, and of course, @Syntax753, I will merge and publish a PR.

**Comment by Syntax753** [comment]
@kochka ! Good to see you! I wasn't sure how active the repo was but just looking into the oAuth requests working and then hopefully submitting a working PR 😄

My background is Java/Maven (4 years ago) but not Android/Gradle - though with your existing work, @matin's OAuth skills, and my typin,g we should get there 👍😄

**Comment by kochka** [comment]
@Syntax753 Good luck with that 😄 If you do not figure out the last bits to make it works, I'll try to find some time during the we to help.

**Comment by jesperuc** [comment]
Would it be possible to Add a paid version/ pro version so we can support the development of the app (actually just this one service - upload to Garmin Connect :-) ) 

looking forward to this will hopefully be fixed one day.

**Comment by Syntax753** [comment]
I got side-tracked since the New Year with work so did not manage to
complete the fix. I am a bit confused by the flow unfortunately :(

**Comment by jesperuc** [comment]
That's completely fair - I think we all now that situation now and then :-) 

Unfortunately I dont have the knowledge on coding to provide my help.

**Comment by Maarty** [comment]
@Syntax753 if you need any help, I can guide you with the flow, unfortunately, I am not very familiar with Java/Android.  
Following the flow I wrote in the comment above should be enough. You need to implement steps 2-4. 
For constructing the Oauth1 header you can find any tutorial on how it should look, or you can use a library - For example here https://oauth.net/1/ are a few libraries also for Java.  
Basically, the header should look something like this
 `"OAuth oauth_consumer_key="{consumerKey}",oauth_signature_method="{signatureMethod}",oauth_timestamp="{timestamp}",oauth_nonce="{nonce}",oauth_version="{version}",oauth_signature="{signature}"`

You should feed the library with the Oauth1 token, consumer key and consumer secret. And it should return you the whole header. Basically, it calculates the signature, which is some HMAC-SHA1 hash created from the provided input.

**Comment by Syntax753** [comment]
Thanks Martin!

I have a branch with what I believe should work but I couldn't quite get it
working. I can try and revisit this when I get some time.

**Comment by louisjennings** [comment]
I got the auth flow working a few days ago (and recently added support for MFA) on a branch based on @Syntax753's work.

My most recent commit is able to successfully upload files to Garmin and I am currently running a debug build on my phone. I need to clean up the code and add support for saving and refreshing authentication. Once that is done, I'd be happy to get the changes merged in.

If you don't have MFA (multi factor authentication/two factor authentication) on your account, the current state of the code should be workable. As I haven't yet added support for OAuth2 refresh, if MFA is enabeld, you'll have to enter an MFA code every time you upload (which could be annoying).

I don't plan on releasing APKs on my forked repo unless I'm unable to get the changes merged into this repo. If you want the changes before the PR is merged, you can build from source.

The branch can be found here: https://github.com/louisjennings/WeightLogger/tree/unable-to-upload-to-garmin-%2368

**Comment by Syntax753** [comment]
Great work!
The project is alive (just niche :)) so a PR should be picked up

**Comment by Syntax753** [comment]
I've reviewed your changes @louisjennings  and they look very good to me!
Can a PR be raised @kochka?

**Comment by kochka** [comment]
@Syntax753 Yes of course !

**Comment by Petr-Mi** [comment]
Any news?

**Comment by atavius** [comment]
@kochka Are there any updates when new version with this fix will be released?

**Comment by kochka** [comment]
I didn't receive any PR, it was a work in progress.

**Comment by atavius** [comment]
Ok I am a bit confused as it seem to me that the fix is available based on comments.
@louisjennings Is your branch working? If so can you create PR for Kochka?

**Comment by Petr-Mi** [comment]
Any news?

**Comment by louisjennings** [comment]
I have updated my branch with changes that fix the upload to Garmin (whilst also adding support for MFA). I've opened a WIP PR (#76) (which will be ready to merge after I have resolved an issue with possible token revokation/invalidation).

I have released a beta build (https://github.com/louisjennings/WeightLogger/releases/tag/v2.5.0-beta.0) and would be keen for some testers to ensure that the changes work across a variety of configurations. The apk is a debug build and likely won't install, except through ADB. I suggest that only developers who are willing to deal with potential issues try this release.

**Comment by louisjennings** [comment]
The known issue with last build is resolved and the PR has been changed to ready to review.
The changelog and latest build can be found here: https://github.com/louisjennings/WeightLogger/releases/tag/v2.5.0-beta.1

**Comment by Syntax753** [comment]
So good to see this being brought to life!
Thanks for the hard work on this :)

**Comment by Petr-Mi** [comment]
> The known issue with last build is resolved and the PR has been changed to ready to review.
> The changelog and latest build can be found here: https://github.com/louisjennings/WeightLogger/releases/tag/v2.5.0-beta.2

No permissions for access to files.
Cannot restore backup and make backup.

**Comment by louisjennings** [comment]
> No permissions for access to files.
> Cannot restore backup and make backup. And save FIT file.

I think this has been fixed now (https://github.com/louisjennings/WeightLogger/commit/61e27eaf6fa4c0c1f19cd185df7f75038553e88c / https://github.com/louisjennings/WeightLogger/releases/tag/v2.5.0-beta.3). Looks like it was an issue arising from targeting newer versions of Android and the associated permissions changes. I didn't pick it up on my local device as I wasn't expecting storage issues and was using the ~/Downloads folder, which was unaffected by the changes.

**Comment by Petr-Mi** [comment]
Beta 3 working now.
Great work Louis!
